nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/realm/certificate/CertificateRealm.java

```diff
@@ -171,80 +129,87 @@ public String getAuthType() {
-     * @exception InvalidOperationException thrown if the realm does not support this operation - e.g. Certificate realm
-     * does not support this operation.
-     * @throws com.sun.enterprise.security.auth.realm.NoSuchUserException
+     * WARN: does not have access to user's certificate, so it does not return groups based on certificate.
      *
+     * @return enumeration of group names assigned to all users authenticated by this realm.
      */
     @Override
-    public Enumeration<String> getGroupNames(String username) throws NoSuchUserException, InvalidOperationException {
-        // This is called during web container role check, not during
-        // EJB container role cheks... fix RI for consistency.
-
-        // Groups for cert users is empty by default unless some assign-groups
-        // property has been specified (see init()).
-        return Collections.enumeration(defaultGroups);
+    public Enumeration<String> getGroupNames(String username) {
+        String[] groups = addAssignGroups(null);
+        if (groups == null) {
+            return Collections.emptyEnumeration();
+        }
+        return Collections.enumeration(Arrays.asList(groups));
     }
 
+
     /**
-     * Returns the name of all the groups that this user belongs to.
-     *
      * @param subject The Subject object for the authentication request.
-     * @param callerPrincipal The Principal object from the user certificate.
-     *
+     * @param principal The Principal object from the user certificate.
+     * @return principal's name
      */
-    public String authenticate(Subject subject, X500Principal callerPrincipal) {
-        // It is important to use X500Principal.getName() as that will
-        // return the LDAP name in RFC2253
-        String callerPrincipalName = callerPrincipal.getName(X500Principal.RFC2253, OID_MAP);
-
-        // Checks if the property for using common name is set
-        if (Boolean.valueOf(getProperty("useCommonName"))) {
-            callerPrincipalName = extractCN(callerPrincipalName);
-        }
-
-        _logger.log(FINEST, "Certificate realm setting up security context for: {0}", callerPrincipalName);
-
-        // Optionally add groups that indicate this caller was authenticated via certificates
-        if (defaultGroups != null) {
-            Set<Principal> principalSet = subject.getPrincipals();
-            for (String groupName : defaultGroups) {
-                principalSet.add(new Group(groupName));
-            }
+    public String authenticate(Subject subject, X500Principal principal) {
+        _logger.finest(() -> String.format("authenticate(subject=%s, principal=%s)", subject, principal));
+
+        final LdapName dn = getLdapName(principal);
+        _logger.log(Level.FINE, "dn={0}", dn);
+        final String principalName = getPrincipalName(dn);
+        _logger.log(Level.FINE, "Certificate realm is setting up security context for principal: '{0}'", principalName);
```

Comment:
Tricky bug ... apostrophes in '{0}' prevent replacement of the value.

Reply:
I'm starting to hate JUL pretty much ... :-(
And I know one way how to speed up anything in Payara :D